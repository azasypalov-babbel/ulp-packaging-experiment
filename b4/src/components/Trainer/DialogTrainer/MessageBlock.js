import React, { useEffect, useRef, useState } from 'react';
import styled, { css } from 'styled-components';
import PropTypes from 'prop-types';
import SpeakerAvatar from './SpeakerAvatar';
import MessageItem from './MessageItem';
import { rem } from 'polished';
import { scrollIntoSafeArea } from '../../shared/scrollIntoSafeArea';


const StyledMessageBlock = styled.div`
  display: flex;
  flex-direction: column;

  cursor: ${(props) => props.onClick ? 'pointer' : 'initial'};

  ${(props) => props.hasAvatar && css`margin-top: ${rem('12px')};`};

  @media (min-width: ${(props) => props.theme.viewports.breakpoints.xsmall}) {
    flex-direction: row;
  }

  & > div:first-child {
    margin: 0 0 ${rem('12px')} 0;
    ${({ position }) => {
    if (position === 'right') return css`
        @media (min-width: ${(props) => props.theme.viewports.breakpoints.xsmall}) {
          margin: 0 0 0 ${rem('12px')};
        }
      `;

    if (position === 'left') return css`
      @media (min-width: ${(props) => props.theme.viewports.breakpoints.xsmall}) {
        margin: 0 ${rem('12px')} 0 0; }`;
  }}
  }
`;

const StyledRow = styled.div`
  display: flex;
  align-items: flex-start;

  &:focus {
    outline: none;
  }

  ${({ position }) => {
    if (position === 'left') return css`
      flex-direction: row;
    `;
    if (position === 'right') return css`
      flex-direction: row-reverse;
      & > div {
        flex-direction: column;
        align-items: flex-end;
      }

      @media (min-width: ${(props) => props.theme.viewports.breakpoints.xsmall}) {
        & > div {
          align-items: flex-start;
          flex-direction: row-reverse;
        }
      }
    `;
  }};
`;

const AvatarSpacer = styled.div`
  height: ${rem('48px')};
  width: ${rem('48px')};
  min-width: ${rem('48px')};
  display: none;
  @media (min-width: ${(props) => props.theme.viewports.breakpoints.xsmall}) {
    display: block;
  }
`;

const MessageBlock = (props) => {
  const {
    hasAvatar,
    speakerRole,
    children,
    secondaryText,
    position,
    isSpeaking,
    onClick,
    avatarIsSpeaking,
    active,
    dataItemType
  } = props;

  const [isSpeakingByClick, setIsSpeakingByClick] = useState(false);

  const handleClick = (event) => {
    setIsSpeakingByClick(true);
    onClick(event);
  };

  useEffect(() => {
    // This a stateful variant of the isSpeaking prop.
    // It will only be true if isSpeaking was triggered by a user click.
    if (isSpeakingByClick && !isSpeaking) {
      setIsSpeakingByClick(false);
    }
  }, [isSpeakingByClick, isSpeaking]);

  const ref = useRef();

  useEffect(() => {
    if (active) {
      scrollIntoSafeArea(ref.current);
    }
  }, [ref, active]);

  return (
    <StyledRow
      ref={ref}
      position={position}
      tabIndex="0"
    >
      <StyledMessageBlock
        active={isSpeakingByClick}
        onClick={onClick ? handleClick : undefined}
        position={position}
        data-selector="item-sound-replay-wrapper"
        hasAvatar={hasAvatar}
      >
        {hasAvatar ? <SpeakerAvatar
          speakerRole={speakerRole}
          isSpeaking={avatarIsSpeaking}
        /> : <AvatarSpacer />}
        <div position={position} data-item-type={dataItemType}>
          <MessageItem
            secondaryText={secondaryText}
            active={isSpeakingByClick}
            position={position}
          >
            {children}
          </MessageItem>
        </div>
      </StyledMessageBlock>
    </StyledRow>
  );
};

MessageBlock.propTypes = {
  position: PropTypes.oneOf(['left', 'right']).isRequired,
  active: PropTypes.bool,
  avatarIsSpeaking: PropTypes.bool,
  isSpeaking: PropTypes.bool,
  hasAvatar: PropTypes.bool,
  children: PropTypes.node,
  speakerRole: PropTypes.oneOf(['m1', 'm2', 'f1', 'f2', 'n1', 'None']),
  secondaryText: PropTypes.string,
  onClick: PropTypes.func,
  dataItemType: PropTypes.string
};

export default MessageBlock;
