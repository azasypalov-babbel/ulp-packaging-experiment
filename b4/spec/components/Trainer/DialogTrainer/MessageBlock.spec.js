import React from 'react';
import { shallow } from 'enzyme';
import MessageBlock from '../../../../src/components/Trainer/DialogTrainer/MessageBlock';

const props = {
  speakerRole: 'n1',
  children: 'learn language',
  secondaryText: 'display language',
  position: 'right',
  onClick: jest.fn(),
  active: false
};

describe('<MessageBlock />', () => {
  test('it renders', () => {
    const wrapper = shallow(<MessageBlock {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});

