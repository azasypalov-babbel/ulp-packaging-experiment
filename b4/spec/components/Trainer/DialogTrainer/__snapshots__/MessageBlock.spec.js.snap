// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MessageBlock /> it renders 1`] = `
<styled.div
  position="right"
  tabIndex="0"
>
  <styled.div
    active={false}
    data-selector="item-sound-replay-wrapper"
    onClick={[Function]}
    position="right"
  >
    <styled.div />
    <div
      position="right"
    >
      <MessageItem
        active={false}
        position="right"
        secondaryText="display language"
      >
        learn language
      </MessageItem>
    </div>
  </styled.div>
</styled.div>
`;
